scalyr_agent/metrics/functions.py

```diff
@@ -216,41 +287,122 @@ def calculate(cls, monitor, metric_name, metric_value, timestamp=None):
+        which takes extra_field values into account is performed when "calculate()" method is called.
         """
         if not monitor or not monitor._global_config:
             return False
 
+        config_calculate_rate_metric_names = (
+            monitor._global_config.calculate_rate_metric_names
+        )
+        monitor_calculate_rate_metric_names = monitor.get_calculate_rate_metric_names()
+
+        # Partial entry key without extra fields suffix
         config_entry_key = "%s:%s" % (monitor.monitor_module_name, metric_name)
 
+        for config_value in chain(
+            config_calculate_rate_metric_names, monitor_calculate_rate_metric_names
+        ):
+            has_extra_fields_filter = config_value.count(":") == 2
+
+            if has_extra_fields_filter:
+                config_value_without_extra_fields_filter = config_value.rsplit(":", 1)[
+                    0
+                ]
+            else:
+                config_value_without_extra_fields_filter = config_value
+
+            # Direct simple metric name only match with no additional extra fields filters
+            if config_value_without_extra_fields_filter == config_entry_key:
+                return True
+
+        return False
+
+    @classmethod
+    def should_calculate(cls, monitor, metric_name, extra_fields=None):
+        """
+        Return True if client side rate should be calculated for the provided metric name and
+        extra_fields values.
+        """
+        extra_fields = extra_fields or {}
+
+        if not monitor or not monitor._global_config:
+            return False
+
         config_calculate_rate_metric_names = (
             monitor._global_config.calculate_rate_metric_names
         )
         monitor_calculate_rate_metric_names = monitor.get_calculate_rate_metric_names()
 
-        return (
-            config_entry_key in config_calculate_rate_metric_names
-            or config_entry_key in monitor_calculate_rate_metric_names
-        )
+        # Config values follow this notation: <monitor module name>:<metric name>:<optional extra field value>
+        # For example: openmetrics_monitor:docker.cpu_usage_seconds_total:mode=kernel
+
+        # Partial entry key without extra fields suffix
+        config_entry_key = "%s:%s" % (monitor.monitor_module_name, metric_name)
+
+        for config_value in chain(
+            config_calculate_rate_metric_names, monitor_calculate_rate_metric_names
+        ):
+            # Direct simple metric name only match
+            if config_value == config_entry_key:
+                return True
+            elif (
+                extra_fields
+                and config_value.count(":") == 2
+                and config_value.count("=") == 1
+            ):
+                config_extra_field_name, config_extra_field_value = config_value.split(
+                    ":"
+                )[-1].split("=")
```

Comment:
Sorry for my lack of knowledge of how metrics are emitted, but isn't it possible to emit a metric with  multiple extra fields? I guess we jus limit configuring rate calculation to one extra field for more simplicity?

Reply:
Metrics can have arbitrary number of extra fields (labels / tags). In fact, it's a very common scenarios for metrics to have multiple labels.

Code for tracking and calculating rates supports metrics with arbitrary number of labels (and each unique label with the same metric name represents a different unique metric and is tracked as such).

This code here just takes care of the configuration where we only allow user to specify a single label (which covers our current use cases).

Let's say we have metrics like this:


```bash
process_cpu_seconds_total app="app-1" node="node-1" mode="user" 155
process_cpu_seconds_total app="app-2" node="node-1" mode="user" 155
process_cpu_seconds_total app="app-3" node="node-1" mode="kernel" 155
process_cpu_seconds_total app="app-3" node="node-3" mode="kernel" 155
```

And the following configuration option is specified: ``openmetrics_monitor:process_cpu_seconds_total:mode=user``.

In this case, rate would only be tracked for metrics where ``mode`` label matches ``user``. If user also wants to track rates for metrics with label ``kernel``, they can define another configuration option which looks like this ``openmetrics_monitor:process_cpu_seconds_total:mode=kernel``.

Reply:
Ah, got it now.